File: rust/crates/mcp/tests/mcp_smoke.rs
use serde_json::json;
use streamfind_rust_core::{MethodRegistry, OperationRegistry};
use streamfind_rust_test_support::{advertised_core_tools, catalogue_entries};

/// Tool names `tools/list` must advertise: core operations plus every
/// exposed domain operation. This is derived from the committed semantic
/// catalogue so it remains stable and adapts automatically to new domains.
fn advertised_tool_names() -> Vec<String> {
    let mut names = advertised_core_tools()
        .into_iter()
        .map(|entry| entry["mcp"]["name"].as_str().unwrap().to_owned())
        .collect::<Vec<_>>();
    names.extend(
        catalogue_entries()
            .into_iter()
            .filter(|entry| {
                entry["kind"] == "operation"
                    && entry["domain"] != "streamfind"
                    && entry["exposed"] == true
            })
            .map(|entry| entry["canonical_id"].as_str().unwrap().to_owned()),
    );
    names
}

/// Assert a `tools/list` payload advertises exactly the expected tool names.
fn assert_advertises(tools: &serde_json::Value, expected: &[String]) {
    let actual_names: Vec<String> = tools
        .as_array()
        .unwrap()
        .iter()
        .map(|tool| tool["name"].as_str().unwrap().to_owned())
        .collect();
    let mut expected = expected.to_vec();
    expected.sort();
    let mut actual = actual_names.clone();
    actual.sort();
    assert_eq!(
        actual, expected,
        "advertised tools differ from the catalogue"
    );
    assert_eq!(
        actual_names.len(),
        expected.len(),
        "duplicate or missing tool names"
    );
    assert!(tools
        .as_array()
        .unwrap()
        .iter()
        .all(|tool| tool["inputSchema"]["type"] == "object"));
}

#[test]
fn supports_initialize_and_tool_listing() {
    let registry = MethodRegistry::default();
    let mut operations = OperationRegistry::default();
    streamfind_rust_mass_spec::register_operations(&mut operations).unwrap();
    let mut session = streamfind_rust_mcp::Session::new(&registry, &operations);
    assert_eq!(
        session.handle(&json!({"id": 1, "method": "initialize"}))["result"]["serverInfo"]["name"],
        "streamfind-rust"
    );
    let tools = session.handle(&json!({"id": 2, "method": "tools/list"}));
    assert_advertises(&tools["result"]["tools"], &advertised_tool_names());
    let initialize = session.handle(&json!({"id": 3, "method": "initialize"}));
    let instructions = initialize["result"]["instructions"]
        .as_str()
        .expect("initialize must provide agent usage instructions");
    for term in [
        "create",
        "describe",
        "tools/list",
        "get_available_methods",
        "connect",
    ] {
        assert!(instructions.contains(term), "instructions missing {term}");
    }
    let metadata = tools["result"]["tools"]
        .as_array()
        .unwrap()
        .iter()
        .find(|tool| tool["name"] == "set_metadata")
        .unwrap();
    assert!(metadata["inputSchema"]["required"]
        .as_array()
        .unwrap()
        .iter()
        .any(|value| value == "metadata"));
}

#[test]
fn method_discovery_returns_complete_input_schemas() {
    let mut registry = MethodRegistry::default();
    streamfind_rust_mass_spec::register_methods(&mut registry).unwrap();
    let operations = OperationRegistry::default();
    let mut session = streamfind_rust_mcp::Session::new(&registry, &operations);
    let response = session.handle(&json!({
        "id": 1,
        "method": "tools/call",
        "params": {
            "name": "get_available_methods",
            "arguments": {"domain": "mass_spec"}
        }
    }));
    assert_ne!(response["result"]["isError"], true, "{response}");
    let methods: serde_json::Value =
        serde_json::from_str(response["result"]["content"][0]["text"].as_str().unwrap()).unwrap();
    let find_features = methods
        .as_array()
        .unwrap()
        .iter()
        .find(|method| method["id"] == "mass_spec.find_features")
        .expect("find_features must be discoverable as a workflow method");
    assert_eq!(find_features["inputSchema"]["type"], "object");
    assert!(find_features["inputSchema"]["properties"].is_object());
    assert!(find_features["parameters"].is_array());
}

#[test]
fn session_lifecycle_rebinds_catalogue() {
    let registry = MethodRegistry::default();
    let mut operations = OperationRegistry::default();
    streamfind_rust_mass_spec::register_operations(&mut operations).unwrap();
    let mut session = streamfind_rust_mcp::Session::new(&registry, &operations);
    let path = streamfind_rust_test_support::tmp_projects_dir()
        .join("streamfind-rust-mcp-lifecycle.duckdb");
    let _ = std::fs::remove_file(&path);
    let call = |session: &mut streamfind_rust_mcp::Session<'_>, id, name, arguments| {
        session.handle(&json!({"id": id, "method": "tools/call", "params": {"name": name, "arguments": arguments}}))
    };
    assert!(
        call(
            &mut session,
            1,
            "create",
            json!({"database_path": path, "domain": "mass_spec"})
        )["result"]["isError"]
            != true
    );
    // Operations are always advertised; connect changes only the method
    // execution context and does not require a tools/list refresh.
    let tools = session.handle(&json!({"id": 3, "method": "tools/list"}))["result"]["tools"]
        .as_array()
        .unwrap()
        .clone();
    assert_advertises(&json!(tools), &advertised_tool_names());
    let eic = tools
        .iter()
        .find(|tool| tool["name"] == "mass_spec.get_raw_spectra_eic")
        .unwrap();
    assert_eq!(eic["inputSchema"]["properties"]["targets"]["type"], "array");
    assert_eq!(
        eic["inputSchema"]["properties"]["targets"]["items"]["type"],
        "object"
    );
    assert_eq!(
        eic["inputSchema"]["properties"]["rt_tolerance"]["type"],
        "number"
    );
    assert_eq!(
        eic["inputSchema"]["properties"]["targets"]["examples"][0][0]["id"],
        "caffeine"
    );
    // The nested target object is part of the advertised JSON Schema, not an
    // opaque generic object. This is what lets clients construct EIC inputs
    // without consulting get_available_methods (which lists methods only).
    for field in ["id", "mass", "mz", "rt", "formula", "SMILES", "InChI"] {
        assert!(
            eic["inputSchema"]["properties"]["targets"]["items"]["properties"]
                .get(field)
                .is_some(),
            "targets schema is missing field {field}"
        );
    }
    let pre_connect_info = call(
        &mut session,
        4,
        "mass_spec.get_analyses_info",
        json!({"database_path": path}),
    );
    assert_ne!(
        pre_connect_info["result"]["isError"], true,
        "{pre_connect_info}"
    );
    let info = call(
        &mut session,
        5,
        "mass_spec.get_analyses_info",
        json!({"database_path": path}),
    );
    assert_ne!(info["result"]["isError"], true, "{info}");
    assert_eq!(
        serde_json::from_str::<serde_json::Value>(
            info["result"]["content"][0]["text"].as_str().unwrap()
        )
        .unwrap(),
        json!({
            "row_count": 0,
            "columns": {
                "analysis": [],
                "replicate": [],
                "blank": [],
                "file_path": [],
                "format": [],
                "number_spectra": [],
                "number_chromatograms": []
            }
        })
    );
    assert!(
        call(&mut session, 5, "close", json!({"database_path": path}))["result"]["isError"] != true
    );
    // After close, operations remain advertised because they are stateless;
    // only the connected method context is cleared.
    let closed =
        session.handle(&json!({"id": 5, "method": "tools/list"}))["result"]["tools"].clone();
    assert_advertises(&closed, &advertised_tool_names());
    let _ = std::fs::remove_file(path);
}

File: cpp/tests/unit/mcp_smoke.cpp
#include <algorithm>
#include <cassert>
#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

#include "streamfind/catalogue.hpp"
#include "streamfind/mcp.hpp"
#include "streamfind/mass_spec/register.hpp"
#include "streamfind/raman/register.hpp"
#include "streamfind/sensors/register.hpp"
#include "../tmp_projects.hpp"

#ifndef STREAMFIND_CATALOGUE_PATH
#error STREAMFIND_CATALOGUE_PATH is required
#endif

namespace {

/// Core tool names `tools/list` must advertise, derived from the committed
/// semantic catalogue (the same file the MCP servers serve from): exposed
/// operations of the `streamfind` domain. Tests adapt automatically when
/// operations are added or removed — no hardcoded counts.
std::vector<std::string> expected_core_tools(const streamfind::Json &entries) {
    std::vector<std::string> names;
    for (const auto &entry : entries) {
        if (entry.value("kind", "") == "operation" &&
            entry.value("domain", "") == "streamfind" &&
            entry.value("exposed", false)) {
            names.push_back(entry.at("mcp").at("name").get<std::string>());
        }
    }
    std::sort(names.begin(), names.end());
    return names;
}

/// The full expected tool set: core tools + all registered domain operations
/// (methods are never tools). Operations are always advertised because they
/// are stateless and carry their own database_path.
std::vector<std::string> expected_all_tools(const streamfind::Json &entries,
                                             const streamfind::OperationRegistry &operations) {
    auto names = expected_core_tools(entries);
    for (const auto &definition : operations.list("")) names.push_back(definition.id);
    std::sort(names.begin(), names.end());
    return names;
}

/// Assert a tools/list payload advertises exactly the expected tool names.
void assert_advertises(const streamfind::Json &tools, const std::vector<std::string> &expected) {
    std::vector<std::string> actual;
    for (const auto &tool : tools) actual.push_back(tool.at("name").get<std::string>());
    std::sort(actual.begin(), actual.end());
    assert(actual == expected);
    assert(tools.size() == expected.size());
    for (const auto &tool : tools) assert(tool.at("inputSchema").at("type") == "object");
}

}  // namespace

int main() {
    const char *phase = "startup";
    try {
    // Load the catalogue explicitly so the test exercises the same artifact the
    // MCP servers query at runtime.
    const auto entries = streamfind::catalogue::load(STREAMFIND_CATALOGUE_PATH);
    assert(entries && "catalogue load failed");
    const auto expected_core = expected_core_tools(*entries);

    phase = "global tools/list";
    const auto response = streamfind::mcp::handle({{"id", 1}, {"method", "tools/list"}});

    const auto tools = response.at("result").at("tools");
    assert_advertises(tools, expected_core);
    phase = "initialize";
    const auto initialize = streamfind::mcp::handle({{"id", 0}, {"method", "initialize"}});
    const auto instructions = initialize.at("result").at("instructions").get<std::string>();
    for (const auto &term : {"create", "describe", "tools/list", "get_available_methods", "connect"}) {
        assert(instructions.find(term) != std::string::npos);
    }
    // A well-known core tool keeps a behavioural contract check: its required
    // parameters are statically meaningful and change only with that operation.
    const auto create = std::find_if(tools.begin(), tools.end(), [](const auto &tool) {
        return tool.at("name") == "create";
    });
    assert(create != tools.end());
    const auto required = create->at("inputSchema").at("required");
    assert(std::find(required.begin(), required.end(), "database_path") != required.end());

    const auto path = streamfind::test::tmp_projects_dir() / "streamfind-mcp-lifecycle.duckdb";
    std::error_code error;
    std::filesystem::remove(path, error);
    streamfind::MethodRegistry registry;
    streamfind::OperationRegistry operations;
    streamfind::mass_spec::register_plugin(*entries, registry, operations);
    streamfind::raman::register_plugin(*entries, registry, operations);
    streamfind::sensors::register_plugin(*entries, registry, operations);
    const auto expected_all = expected_all_tools(*entries, operations);
    streamfind::mcp::Session session(registry, operations);
    phase = "get_available_methods";
    const auto methods_call = session.handle({{"id", 9}, {"method", "tools/call"}, {"params", {
        {"name", "get_available_methods"}, {"arguments", {{"domain", "mass_spec"}}}
    }}});
    const auto methods = streamfind::Json::parse(methods_call.at("result").at("content").at(0).at("text").get<std::string>());
    const auto find_features = std::find_if(methods.begin(), methods.end(), [](const auto &method) {
        return method.at("id") == "mass_spec.find_features";
    });
    assert(find_features != methods.end());
    assert(find_features->at("inputSchema").at("type") == "object");
    assert(find_features->at("inputSchema").at("properties").is_object());

    phase = "create";
    const auto create_call = session.handle({{"id", 2}, {"method", "tools/call"}, {"params", {
        {"name", "create"}, {"arguments", {{"database_path", path.string()}, {"domain", "mass_spec"}}}
    }}});
    if (create_call.at("result").value("isError", false)) {
        std::cerr << "create failed\n";
        return 1;
    }
    const auto before_connect = session.handle({{"id", 3}, {"method", "tools/list"}}).at("result").at("tools");
    assert_advertises(before_connect, expected_all);
    for (const auto &operation : {"raman.add_analyses", "raman.remove_analyses"}) {
        const auto advertised = std::find_if(before_connect.begin(), before_connect.end(),
                                             [operation](const auto &tool) {
                                                 return tool.at("name") == operation;
                                             });
        assert(advertised != before_connect.end());
        assert(advertised->at("inputSchema").at("type") == "object");
    }
    const auto sensors_tool = std::find_if(before_connect.begin(), before_connect.end(), [](const auto &tool) {
        return tool.at("name").template get<std::string>().rfind("sensors.", 0) == 0;
    });
    assert(sensors_tool == before_connect.end());
    const auto eic_before_connect = std::find_if(before_connect.begin(), before_connect.end(), [](const auto &tool) {
        return tool.at("name") == "mass_spec.get_raw_spectra_eic";
    });
    assert(eic_before_connect != before_connect.end());
    assert(eic_before_connect->at("inputSchema").at("properties").at("targets").at("items").contains("properties"));
    const auto pre_connect_info = session.handle({{"id", 10}, {"method", "tools/call"}, {"params", {
        {"name", "mass_spec.get_analyses_info"}, {"arguments", {{"database_path", path.string()}}}
    }}});
    assert(!pre_connect_info.at("result").value("isError", false));
    phase = "connect";
    const auto connect = session.handle({{"id", 4}, {"method", "tools/call"}, {"params", {
        {"name", "connect"}, {"arguments", {{"database_path", path.string()}}}
    }}});
    if (connect.at("result").value("isError", false)) {
        std::cerr << "connect failed\n";
        return 1;
    }
    const auto connected_tools = session.handle({{"id", 5}, {"method", "tools/list"}}).at("result").at("tools");
        assert_advertises(connected_tools, expected_all);
    const auto eic_tool = std::find_if(connected_tools.begin(), connected_tools.end(), [](const auto &tool) {
        return tool.at("name") == "mass_spec.get_raw_spectra_eic";
    });
    if (eic_tool == connected_tools.end()) {
        std::cerr << "mass-spec target schema mismatch: EIC tool absent\n";
        return 1;
    }
    const auto &eic_schema = eic_tool->at("inputSchema");
    const auto &targets_schema = eic_schema.at("properties").at("targets");
    const auto &target_item_schema = targets_schema.at("items");
    const bool schema_ok = targets_schema.at("type") == "array" &&
                           target_item_schema.at("type") == "object" &&
                           eic_schema.at("properties").at("rt_tolerance").at("type") == "number" &&
                           !targets_schema.at("examples").empty() &&
                           targets_schema.at("examples")[0][0].at("id") == "caffeine" &&
                           target_item_schema.contains("properties") &&
                           target_item_schema.at("properties").contains("mz");
    if (!schema_ok) {
        std::cerr << "mass-spec target schema mismatch: " << eic_schema.dump() << '\n';
        return 1;
    }
    const auto info = session.handle({{"id", 5}, {"method", "tools/call"}, {"params", {
        {"name", "mass_spec.get_analyses_info"}, {"arguments", {{"database_path", path.string()}}}
    }}});
    if (info.at("result").value("isError", false) ||
        streamfind::Json::parse(info.at("result").at("content").at(0).at("text").get<std::string>()).at("row_count") != 0) {
        std::cerr << "mass_spec.get_analyses_info failed\n";
        return 1;
    }
    const auto close = session.handle({{"id", 6}, {"method", "tools/call"}, {"params", {
        {"name", "close"}, {"arguments", {{"database_path", path.string()}}}
    }}});
    if (close.at("result").value("isError", false)) {
        std::cerr << "close failed\n";
        return 1;
    }
    assert_advertises(session.handle({{"id", 8}, {"method", "tools/list"}}).at("result").at("tools"), expected_all);
    std::filesystem::remove(path, error);
    } catch (const std::exception &error) {
        std::cerr << "mcp_smoke exception during " << phase << ": " << error.what() << '\n';
        return 1;
    }
}